### PR TITLE
✨ [Extension AMP-AUTO-ADS] Implement feature custom amp-analytics by amp-auto-ads

### LINKED ISSUE #30997: amp-analytics doesn't find an element
## What's the issue?

I am using `amp-analytics` to track my ads, but when using `amp-analytic` with `amp-auto-ads` it doesn’t track for the elements created after the `amp-analytics` has been loaded.

I believe this is happening with all amp extensions that dynamically loads elements on the page and not only `amp-auto-ads`.

Error message

> Element "XXX" not found:  null

## How do we reproduce the issue?

I created 2 environments trying to simulate a delay on the extension so I can reproduce the error that occurs depending on user’s network:

1 – I delayed in 1 second the `amp-analytics` request: https://delay-amp-analytics.denakop.com/amp.html

2 – I delayed in 1 second the `amp-auto-ad` request: https://delay-amp-auto-ads.denakop.com/amp.html

When delaying the `amp-analytics` the error will not occur because the selector exists.

When delaying `amp-auto-ad` the `amp-analytics` will throw an error because when it is loaded the `<amp-ad>` elements is not created yet

## What browsers are affected?

I managed to reproduce the error on all major browsers and devices.

## Which AMP version is affected?

I believe this issue is happening in all AMP versions

### PR DESCRIPTION
**Pull request by kelvinbiffi**
The feature on this PR is to be able to add a custom amp-analytics by amp-auto-ads to be able to collecting the ad events more precisally.

This functionality is so important to  companies who need track real time ad data information.

Closes #30997

### CODE REVIEW

**Comment by kelvinbiffi** [comment]
Hi @calebcordry I'd added the unit tests and follow the CI requirements, could you please have a look on this PR? =)

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js [hunk 1]
```diff
@@ -70,6 +74,22 @@ export class AnchorAdStrategy {
...
+   */
+  addCustomAmpAnalytics_(adElement) {
+    if (this.customAnalytics_) {
+      const doc = this.ampdoc.win.document;
+      const customAmpAnalytics = createElementWithAttributes(
+        doc,
+        'amp-analytics',
+        this.customAnalytics_
+      );
+      adElement.appendChild(customAmpAnalytics);
```
Comment (alanorozco) [inline extensions/amp-auto-ads/0.1/anchor-ad-strategy.js:89]: Extensions are not allowed to create analytics this way. Please use `insertAnalyticsElement` from `src/extension-analytics` instead. This sandboxes the inserted element for user privacy.

**Review by alanorozco — CHANGES_REQUESTED** [review]

File: extensions/amp-auto-ads/0.1/amp-auto-ads.js [hunk 2]
```diff
@@ -33,6 +33,11 @@ export class AmpAutoAds extends AMP.BaseElement {
     this.adNetwork_ = getAdNetworkConfig(type, this.element);
     userAssert(this.adNetwork_, 'No AdNetworkConfig for type: ' + type);
 
+    this.customAnalytics_ = this.element.getAttribute('custom-analytics');
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/amp-auto-ads.js:36]: nit: 
```js
if(this.element.hasAttribute('custom-analytics')) {
  this.customAnalytics_ = ...
}
```

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js [hunk 3]
```diff
@@ -15,8 +15,9 @@ export class AnchorAdStrategy {
    * @param {!JsonObject<string, string>} baseAttributes Any attributes that
    *     should be added to any inserted ads.
    * @param {!JsonObject} configObj
+   * @param {!JsonObject} customAnalytics
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/anchor-ad-strategy.js:18]: can be null here

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js [hunk 4]
```diff
@@ -25,6 +26,9 @@ export class AnchorAdStrategy {
 
     /** @const @private {!JsonObject} */
     this.configObj_ = configObj;
+
+    /** @const @private {!JsonObject} */
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/anchor-ad-strategy.js:30]: can be null

File: extensions/amp-auto-ads/0.1/placement.js [hunk 5]
```diff
@@ -91,14 +91,16 @@ export class Placement {
...
+   * @param {!JsonObject} customAnalytics
    */
   constructor(
     ampdoc,
     anchorElement,
     position,
     injector,
     attributes,
-    opt_margins
+    opt_margins,
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/placement.js:102]: optional param must be last

File: extensions/amp-auto-ads/0.1/placement.js [hunk 6]
```diff
@@ -91,14 +91,16 @@ export class Placement {
    * @param {function(!Element, !Element)} injector
    * @param {!JsonObject<string, string>} attributes
    * @param {!../../../src/layout-rect.LayoutMarginsChangeDef=} opt_margins
+   * @param {!JsonObject} customAnalytics
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/placement.js:94]: can be null

File: extensions/amp-auto-ads/0.1/placement.js [hunk 7]
```diff
@@ -169,6 +174,21 @@ export class Placement {
     }
   }
 
+  /**
+   * Add custom amp-analytics emement
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/placement.js:178]: nit element.

File: extensions/amp-auto-ads/0.1/placement.js [hunk 8]
```diff
@@ -129,6 +131,9 @@ export class Placement {
 
     /** @private {!PlacementState} */
     this.state_ = PlacementState.UNUSED;
+
+    /** @const @private {!JsonObject} */
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/placement.js:135]: nullable

File: extensions/amp-auto-ads/0.1/placement.js [hunk 9]
```diff
@@ -169,6 +174,21 @@ export class Placement {
...
+   */
+  addCustomAmpAnalytics_() {
+    if (this.customAnalytics_) {
+      const doc = this.ampdoc.win.document;
+      const customAmpAnalytics = createElementWithAttributes(
+        doc,
+        'amp-analytics',
+        this.customAnalytics_
+      );
+      this.getAdElement().appendChild(customAmpAnalytics);
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/placement.js:188]: why do we need to create both here and in anchor-ad-strategy?

**Review by calebcordry — CHANGES_REQUESTED** [review]

File: extensions/amp-auto-ads/0.1/placement.js (hunk 9 above)
Comment (kelvinbiffi) [reply extensions/amp-auto-ads/0.1/placement.js:188]: @calebcordry When I implement the feature I've seen different flows, there is no mention of 'placement' inside 'ad-anchor-strategy', so I see the needed of add the amp-analytics in both, to the amp-analytics be added inside the 'amp-ad' properly

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js (hunk 1 above)
Comment (kelvinbiffi) [reply extensions/amp-auto-ads/0.1/anchor-ad-strategy.js:89]: @alanorozco I tried to use the `insertAnalyticsElement` to create a `amp-analytics`, however thereare no way of create a `amp-analytics` with url config, only with script config pattern, so I made a modification and added a optional `opt_useUrlConfig` parameter to `insertAnalyticsElement` function to be possible to create a  amp-analytics with url config pattern

**Comment by kelvinbiffi** [comment]
@calebcordry @alanorozco I'd done the fixes, could you take a look? =)

psc @gurodrigues

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js (hunk 1 above)
Comment (kelvinbiffi) [reply extensions/amp-auto-ads/0.1/anchor-ad-strategy.js]: @alanorozco I changed the logic to fetch the configuration from the url config to keep the sandbox flow

**Comment by kelvinbiffi** [comment]
@alanorozco @calebcordry I saw in amp source code that is not possible creating an amp-analytics with config by an URL and SANDBOX mode in the same time, so I created the flux of create the amp-analytics with url config without the sandbox mode, could we go forward with this solution?

File: extensions/amp-auto-ads/0.1/anchor-ad-strategy.js [hunk 10]
```diff
@@ -70,6 +75,22 @@ export class AnchorAdStrategy {
...
+   * @param {HTMLElement} adElement
+   */
+  addCustomAmpAnalytics_(adElement) {
+    if (this.customAnalytics_) {
+      insertAnalyticsElement(
+        adElement,
+        this.customAnalytics_,
+        false,
+        false,
+        true
```
Comment (alanorozco) [inline extensions/amp-auto-ads/0.1/anchor-ad-strategy.js:89]: Please annotate bool args.

```suggestion
        /* config */ false,
        /* loadAnalytics */ false,
        /* disableImmediate */ true
```

File: extensions/amp-auto-ads/0.1/amp-auto-ads.js [hunk 11]
```diff
@@ -114,7 +119,11 @@ export class AmpAutoAds extends AMP.BaseElement {
         this.user().warn(TAG, noConfigReason);
       }
 
-      const placements = getPlacementsFromConfigObj(ampdoc, configObj);
+      const placements = getPlacementsFromConfigObj(
+        ampdoc,
+        configObj,
+        this.customAnalytics_
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/amp-auto-ads.js:125]: lets make sure we annotate this as nullable the whole way down.

File: extensions/amp-auto-ads/0.1/amp-auto-ads.js [hunk 12]
```diff
@@ -33,6 +33,11 @@ export class AmpAutoAds extends AMP.BaseElement {
     this.adNetwork_ = getAdNetworkConfig(type, this.element);
     userAssert(this.adNetwork_, 'No AdNetworkConfig for type: ' + type);
 
+    if (this.element.hasAttribute('custom-analytics')) {
+      this.customAnalytics_ = this.element.getAttribute('custom-analytics');
+      this.customAnalytics_ = JSON.parse(this.customAnalytics_);
```
Comment (calebcordry) [inline extensions/amp-auto-ads/0.1/amp-auto-ads.js:38]:
```suggestion
      const customAnalytics = this.element.getAttribute('custom-analytics');
      this.customAnalytics_ = JSON.parse(customAnalytics);
```

**Review by calebcordry — APPROVED** [review]
Ads changes LGTM with a couple nits. @alanorozco just to make sure you are good with analytics api change

**Comment by kelvinbiffi** [comment]
@calebcordry @alanorozco suggestions and fixeds done =)

**Comment by kelvinbiffi** [comment]
Hi @alanorozco ,

Did you get a look?

Is there anything I need to do?

**Comment by kelvinbiffi** [comment]
Hi @alanorozco 

I made on this way to avoid fetch the remote config because it would make the function asynchronous

**Comment by kelvinbiffi** [comment]
Hi @alanorozco ,

Did you get a look?

Is there anything I need to do?

**Comment by kelvinbiffi** [comment]
Hi @alanorozco @calebcordry ,

Did you get a look?

Is there anything I need to do?

**Comment by douglaspoa** [comment]
hello @alanorozco What is still missing to be approved?

**Comment by douglaspoa** [comment]
Hello @alanorozco , can you give us an update?

**Pull request closed without merging**